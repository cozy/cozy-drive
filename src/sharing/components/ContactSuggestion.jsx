import React from 'react'
import PropTypes from 'prop-types'
import get from 'lodash/get'
import { Avatar } from 'cozy-ui/react'
import { translate } from 'cozy-ui/react/I18n'

import styles from 'sharing/components/recipient.styl'

import { Contact, Group } from 'models'
import Identity from 'sharing/components/Identity'

export const ContactSuggestion = ({ contactOrGroup, contacts, t }) => {
  let avatarText, name, details
  if (contactOrGroup._type === Group.doctype) {
    name = contactOrGroup.name
    const membersCount = contacts
      .reduce((total, contact) => {
        if (
          get(contact, 'relationships.groups.data', [])
            .map(group => group._id)
            .includes(contactOrGroup._id)
        ) {
          return total + 1
        }

        return total
      }, 0)
      .toString()
    details = t('Share.members.count', {
      smart_count: membersCount
    })
    avatarText = 'G'
  } else {
    name = Contact.getDisplayName(contactOrGroup)
    avatarText = Contact.getInitials(contactOrGroup)
    details = Contact.getPrimaryCozy(contactOrGroup)
  }

  return (
    <div className={styles['recipient']}>
      <Avatar text={avatarText} size="small" />
      <Identity name={name} details={details} />
    </div>
  )
}

ContactSuggestion.propTypes = {
  contactOrGroup: PropTypes.oneOfType([Contact.propType, Group.propType])
    .isRequired,
  contacts: PropTypes.arrayOf(Contact.propType),
  t: PropTypes.func.isRequired
}

export default translate()(ContactSuggestion)
