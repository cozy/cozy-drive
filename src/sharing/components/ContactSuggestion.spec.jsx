import React from 'react'
import { shallow } from 'enzyme'

import { ContactSuggestion } from './ContactSuggestion'

describe('ContactSuggestion component', () => {
  const fakeT = jest.fn((s, args) => {
    if (s === 'Share.members.count') {
      return `${args.smart_count} members`
    }

    return s
  })
  it('should display contact suggestion for a contact', () => {
    const jonSnow = {
      _id: 'f3a4e501-abbd',
      fullname: 'Jon Snow',
      name: {
        givenName: 'Jon',
        familyName: 'Snow'
      },
      cozy: [
        {
          primary: true,
          url: 'https://jonsnow.mycozy.cloud'
        }
      ],
      email: [
        {
          address: 'jon.snow@email.com',
          type: 'primary'
        }
      ],
      _type: 'io.cozy.contacts'
    }
    const contacts = [jonSnow]
    const props = {
      contactOrGroup: jonSnow,
      contacts,
      t: fakeT
    }
    const jsx = <ContactSuggestion {...props} />
    const wrapper = shallow(jsx)
    expect(wrapper).toMatchSnapshot()
  })

  it('should display contact suggestion for a contact without fullname (for example created by a share modal)', () => {
    const jonSnow = {
      _id: 'f3a4e501-abbd',
      cozy: [],
      email: [
        {
          address: 'jon.snow@email.com',
          type: 'primary'
        }
      ],
      name: undefined,
      _type: 'io.cozy.contacts'
    }
    const contacts = [jonSnow]
    const props = {
      contactOrGroup: jonSnow,
      contacts,
      t: fakeT
    }
    const jsx = <ContactSuggestion {...props} />
    const wrapper = shallow(jsx)
    expect(wrapper).toMatchSnapshot()
  })

  it('should display contact suggestion for a group', () => {
    const jon = {
      _id: 'f3a4e501-abbd',
      fullname: 'Jon Snow',
      name: {
        givenName: 'Jon',
        familyName: 'Snow'
      },
      email: [
        {
          address: 'jon.snow@email.com',
          type: 'primary'
        }
      ],
      _type: 'io.cozy.contacts',
      relationships: {
        groups: {
          data: [{ _id: '610718e6-2d7a' }]
        }
      }
    }
    const cersei = {
      _id: '42dc490a-7878',
      fullname: 'Cersei Lannister',
      name: {
        givenName: 'Cersei',
        familyName: 'Lannister'
      },
      email: [
        {
          address: 'cersei@email.com',
          type: 'primary'
        }
      ],
      _type: 'io.cozy.contacts',
      relationships: { groups: { data: [] } }
    }
    const sam = {
      _id: 'b5bed853-93da',
      fullname: 'Samwell Tarly',
      name: {
        givenName: 'Samwell',
        familyName: 'Tarly'
      },
      email: [
        {
          address: 'samwell@email.com',
          type: 'primary'
        }
      ],
      _type: 'io.cozy.contacts',
      relationships: {
        groups: {
          data: [{ _id: '610718e6-2d7a' }]
        }
      }
    }
    const contacts = [jon, cersei, sam]
    const theNightsWatch = {
      _id: '610718e6-2d7a',
      name: "The Night's Watch",
      _type: 'io.cozy.contacts.groups'
    }
    const props = {
      contactOrGroup: theNightsWatch,
      contacts,
      t: fakeT
    }
    const jsx = <ContactSuggestion {...props} />
    const wrapper = shallow(jsx)
    expect(wrapper).toMatchSnapshot()
  })
})
