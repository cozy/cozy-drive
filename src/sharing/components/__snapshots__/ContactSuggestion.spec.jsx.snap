// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactSuggestion component should display contact suggestion for a contact 1`] = `
<div
  className="recipient"
>
  <Avatar
    className=""
    image=""
    size="small"
    text="JS"
  />
  <Identity
    details="https://jonsnow.mycozy.cloud"
    name="Jon Snow"
  />
</div>
`;

exports[`ContactSuggestion component should display contact suggestion for a contact without fullname (for example created by a share modal) 1`] = `
<div
  className="recipient"
>
  <Avatar
    className=""
    image=""
    size="small"
    text="J"
  />
  <Identity
    details="-"
    name="jon.snow@email.com"
  />
</div>
`;

exports[`ContactSuggestion component should display contact suggestion for a group 1`] = `
<div
  className="recipient"
>
  <Avatar
    className=""
    image=""
    size="small"
    text="G"
  />
  <Identity
    details="2 members"
    name="The Night's Watch"
  />
</div>
`;
